### PR TITLE
adding replication alerting

### PR DESCRIPTION
**Pull request by wejdross**
## Checklist

- [x] The PR has a meaningful title. It will be used to auto generate the
      changelog.
      The PR has a meaningful description that sums up the change. It will be
      linked in the changelog.
- [x] PR contains a single logical change (to build a better changelog).
- [x] Categorize the PR by adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.
- [x] Link this PR to related issues or PRs.

### CODE REVIEW

File: component/vshn_postgres.jsonnet [hunk 1]
```diff
@@ -776,6 +778,65 @@ local prometheusRule = {
...
+              },
+              // new
+              {
+                name: 'postgresql-replication',
+                rules: [
+                  {
+                    alert: 'PostgreSQLReplicationCritical',
+                    annotations: {
+                      description: 'The number of replicas for the instance {{ $labels.cluster_name }} in namespace {{ $labels.namespace }}. Please check pod counts in affected namespace.',
+                      // runbook_url: 'TBD',
```
Comment (Kidswiss) [inline component/vshn_postgres.jsonnet:794]: We probably want to have some runbooks here.

File: component/vshn_postgres.jsonnet [hunk 2]
```diff
@@ -776,6 +778,65 @@ local prometheusRule = {
...
+                ],
+              },
+              {
+                name: 'postgresql-replication-lag',
+                rules: [
+                  {
+                    alert: 'PostgreSQLReplicationLagCritical',
+                    annotations: {
+                      description: 'Replication lag size on namespace {{$labels.exported_namespace}} instance ({{$labels.application_name}}) is currently {{ $value | humanize1024}}B behind the leader.',
+                      // runbook_url: 'TBD',
```
Comment (Kidswiss) [inline component/vshn_postgres.jsonnet:813]: here as well.

File: component/vshn_postgres.jsonnet [hunk 3]
```diff
@@ -776,6 +778,65 @@ local prometheusRule = {
...
+                ],
+              },
+              {
+                name: 'postgresql-replication-count',
+                rules: [
+                  {
+                    alert: 'PostgreSQLReplicationCountCritical',
+                    annotations: {
+                      description: 'Replication is broken in namespace {{$labels.namespace}}, check statefulset ({{$labels.statefulset}}).',
+                      // runbook_url: 'TBD',
```
Comment (Kidswiss) [inline component/vshn_postgres.jsonnet:832]: Also here.

**Review by Kidswiss — CHANGES_REQUESTED** [review]
As these are not so trivial alerts, we probably want some runbooks here. E.g. how to re-initialize the replication via `patronictl`. Or how to figure out the most likely issue that prevents the pods from starting (like the quota case you found), etc.

**Comment by wejdross** [comment]
@Kidswiss I tried to add some general tips how to debug, but since I never fix that in prod it's hard to guess what could be usefull, there are such docs: https://bootvar.com/useful-patroni-commands/ , but I don't see sense in putting it into our docs. wdyt? have You got anything better?

File: component/vshn_postgres.jsonnet [hunk 4]
```diff
@@ -776,6 +778,65 @@ local prometheusRule = {
...
+                    },
+                  },
+                ],
+              },
+              // new
+              {
+                name: 'postgresql-replication',
+                rules: [
+                  {
+                    alert: 'PostgreSQLReplicationCritical',
```
Comment (Kidswiss) [inline component/vshn_postgres.jsonnet:791]:
```suggestion
                    alert: 'PostgreSQLPodReplicasCritical',
```

I feel that makes it clearer what exactly is not working.

**Comment by Kidswiss** [comment]
@wejdross I would explicitly put in the `patroni reinit` for the replication alerts, because that's how you fix a broken replication.

Also for the pod replicas I would add a few commands how to check the statefulset for events.

It doesn't have to be much just a few lines.

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 5]
```diff
@@ -123,3 +123,54 @@ appcat:
...
+(inside pod):bash-4.4$ patronictl list
+(inside pod):bash-4.4$ patronictl reinit test-cluster-always-true-jnlj4 test-cluster-always-true-jnlj4-1
+```
+
+notice that patronictl takes the name of the cluster and the name of the pod, not the name of the statefulset. Can be used in scripts with `--force` flag which auto-accepts all prompts.
+
+if that doesn;t work:
+delete replica pods and let patroni fix itself
+if that doesn't work either, please refer to the patroni documentation on how to fix replication issues.
+https://patroni.readthedocs.io/en/latest/rest_api.html?highlight=reinit#reinitialize-endpoint[Patroni documentation - reinitialize replica pod]
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:152]: Not sure if the rest-api docs help with fixing problems.

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 6]
```diff
@@ -123,3 +123,54 @@ appcat:
...
+https://stackgres.io/doc/1.1/administration/patroni/management/[Stackgres documentation]
+https://www.percona.com/blog/how-patroni-addresses-the-problem-of-the-logical-replication-slot-failover-in-a-postgresql-cluster/[Percona blog post]
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:154]: Here as well..

File: component/vshn_postgres.jsonnet [hunk 7]
```diff
@@ -776,6 +778,65 @@ local prometheusRule = {
...
+                    },
+                  },
+                ],
+              },
+              // new
+              {
+                name: 'postgresql-replication',
+                rules: [
+                  {
+                    alert: 'PostgreSQLPodReplicasCritical',
```
Comment (Kidswiss) [inline component/vshn_postgres.jsonnet:791]: I mentioned in the chat already, but I previously commented here on the wrong line 🙈 

The name for that one was good, but the one on line 829 should be called `PostgreSQLPodReplicasCritical`.

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 8]
```diff
@@ -123,3 +123,54 @@ appcat:
...
+
+We need to build know-how on how to fix replication issues, please document your findings in this runbook.
+
+[[PostgreSQLReplicationLagCritical]]
+== PostgreSQLReplicationLagCritical
+
+This alert fires if the replication lag is higher than 5 minutes for a VSHNPostrgeSQL instance. It means that replicas are behind master.
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:161]:
```suggestion
This alert fires if the replication lag is higher than 5 minutes for a VSHNPostrgeSQL instance. It means that replicas are behind master. If storage and network looks okay, you can try to re-initialize the node with the replication lag.
```

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc (hunk 6 above)
Comment (wejdross) [reply docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc]: deleted

File: component/vshn_postgres.jsonnet (hunk 7 above)
Comment (wejdross) [reply component/vshn_postgres.jsonnet]: ok, restored that

File: docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc [hunk 9]
```diff
@@ -123,3 +123,52 @@ appcat:
...
+We need to build know-how on how to fix replication issues, please document your findings in this runbook.
+
+[[PostgreSQLReplicationLagCritical]]
+== PostgreSQLReplicationLagCritical
+
+This alert fires if the replication lag is higher than 5 minutes for a VSHNPostrgeSQL instance. It means that replicas are behind master. If storage and network looks okay, you can try to re-initialize the node with the replication lag.
+Probably related to network or storage.
+
+
+[[PostgreSQLReplicationCountCritical]]
```
Comment (Kidswiss) [inline docs/modules/ROOT/pages/runbooks/vshn-postgresql.adoc:163]: Then you also need to change it here or the link will point nowhere

**Review by Kidswiss — APPROVED** [review]
Should be a good baseline to get started now.

**Pull request merged**